File: plugin/src/main/java/org/jukeboxmc/plugin/perms/repository/PermissionsRepository.java
package org.jukeboxmc.plugin.perms.repository;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import lombok.RequiredArgsConstructor;
import org.jukeboxmc.plugin.perms.util.MySQL;
import org.redisson.api.RMap;
import org.redisson.api.RedissonClient;

/**
 * @author Kaooot
 * @version 1.0
 */
@Singleton
@RequiredArgsConstructor(onConstructor_ = @Inject)
public class PermissionsRepository {

    private final MySQL mySQL;
    private final RedissonClient redissonClient;

    private RMap<UUID, Set<String>> userPermissions;

    public void setup() {
        try (final PreparedStatement createTable = this.mySQL.prepare("CREATE TABLE IF NOT " +
            "EXISTS permissions(player VARCHAR(36) NOT NULL, permissions LONGTEXT NOT NULL, " +
            "PRIMARY KEY(player))")) {
            createTable.execute();
        } catch (SQLException e) {
            e.printStackTrace();
        }

        this.userPermissions = this.redissonClient.getMap("jperms:user-perms");
    }

    public void insert(UUID player) {
        this.select(player).whenComplete((permissions, throwable) -> {
            if (permissions == null) {
                this.mySQL.prepareAsync("INSERT INTO permissions(player, permissions) VALUES(?,?)")
                    .whenComplete((statement, t) -> {
                        try (statement) {
                            statement.setString(1, player.toString());
                            statement.setString(2, "");
                            statement.execute();
                        } catch (SQLException e) {
                            e.printStackTrace();
                        }
                    });

                this.userPermissions.fastPutAsync(player, new HashSet<>());
            } else {
                this.userPermissions.fastPutAsync(player, permissions);
            }
        });
    }

    public void update(UUID player, Set<String> permissions) {
        this.userPermissions.fastPutAsync(player, permissions);

        this.mySQL.prepareAsync("UPDATE permissions SET permissions = ? WHERE player = ?")
            .whenComplete((statement, throwable) -> {
                try (statement) {
                    statement.setString(1, String.join(", ", permissions));
                    statement.setString(2, player.toString());
                    statement.execute();
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            });
    }

    public CompletableFuture<Set<String>> select(UUID player) {
        return this.mySQL.prepareAsync("SELECT permissions FROM permissions WHERE player = ?")
            .thenApply(statement -> {
                try (statement) {
                    statement.setString(1, player.toString());

                    final ResultSet resultSet = statement.executeQuery();
                    final Set<String> permissions;

                    if (resultSet.next()) {
                        final String s = resultSet.getString("permissions");

                        if (s.contains(",")) {
                            permissions = new HashSet<>(Arrays.stream(s.split(", ")).toList());
                        } else {
                            permissions = new HashSet<>();

                            if (!s.isEmpty()) {
                                permissions.add(s);
                            }
                        }
                    } else {
                        permissions = null;
                    }

                    resultSet.close();

                    return permissions;
                } catch (SQLException e) {
                    e.printStackTrace();

                    return null;
                }
            });
    }

    public void delete(UUID player) {
        this.clearCache(player);

        this.mySQL.prepareAsync("DELETE FROM permissions WHERE player = ?")
            .whenComplete((statement, throwable) -> {
                try (statement) {
                    statement.setString(1, player.toString());
                    statement.execute();
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            });
    }

    public void clearCache(UUID player) {
        this.userPermissions.fastRemoveAsync(player);
    }

    public Set<String> cachedPermissions(UUID player){
        return this.userPermissions.getOrDefault(player, new HashSet<>());
    }

    public boolean hasCachedPerm(UUID player, String perm) {
        return this.userPermissions.getOrDefault(player, new HashSet<>()).contains(perm);
    }
}

File: plugin/src/main/kotlin/org/jukeboxmc/plugin/perms/repository/PermissionsRepository.kt
package org.jukeboxmc.plugin.perms.repository

import org.jukeboxmc.plugin.perms.util.MySQL
import org.redisson.api.RMap
import org.redisson.api.RedissonClient
import java.sql.PreparedStatement
import java.sql.SQLException
import java.util.*
import java.util.concurrent.CompletableFuture

/**
 * @author Kaooot
 * @version 1.0
 */
class PermissionsRepository(private val mySQL: MySQL, private val redissonClient: RedissonClient) {

    private lateinit var userPermissions: RMap<UUID, Set<String>>

    fun setup() {
        try {
            this.mySQL.prepare("CREATE TABLE IF NOT EXISTS permissions(player VARCHAR(36) NOT NULL, permissions LONGTEXT NOT NULL, PRIMARY KEY(player))").use { createTable -> createTable!!.execute() }
        } catch (e: SQLException) {
            e.printStackTrace()
        }

        this.userPermissions = this.redissonClient.getMap("jperms:user-perms")
    }

    fun insert(player: UUID) {
        this.select(player).whenComplete { permissions: Set<String>?, _: Throwable? ->
            if (permissions == null) {
                this.mySQL.prepareAsync("INSERT INTO permissions(player, permissions) VALUES(?,?)").whenComplete { statement: PreparedStatement, _: Throwable? ->
                    try {
                        statement.use {
                            statement.setString(1, player.toString())
                            statement.setString(2, "")
                            statement.execute()
                        }
                    } catch (e: SQLException) {
                        e.printStackTrace()
                    }
                }

                this.userPermissions.fastPutAsync(player, HashSet())
            } else {

                this.userPermissions.fastPutAsync(player, permissions)
            }
        }
    }

    fun update(player: UUID, permissions: MutableSet<String>) {
        this.userPermissions.fastPutAsync(player, permissions)

        this.mySQL.prepareAsync("UPDATE permissions SET permissions = ? WHERE player = ?").whenComplete { statement: PreparedStatement, _: Throwable? ->
            try {
                statement.use {
                    statement.setString(1, java.lang.String.join(", ", permissions))
                    statement.setString(2, player.toString())
                    statement.execute()
                }
            } catch (e: SQLException) {
                e.printStackTrace()
            }
        }
    }

    fun select(player: UUID): CompletableFuture<MutableSet<String>?> {
        return this.mySQL.prepareAsync("SELECT permissions FROM permissions WHERE player = ?").thenApply { statement: PreparedStatement ->
            try {
                statement.use {
                    statement.setString(1, player.toString())

                    val resultSet = statement.executeQuery()
                    val permissions: MutableSet<String>?

                    if (resultSet.next()) {
                        val s = resultSet.getString("permissions")

                        if (s.contains(",")) {
                            permissions = HashSet(s.split(", "))
                        } else {
                            permissions = HashSet()

                            if (s.isNotEmpty()) {
                                permissions.add(s)
                            }
                        }

                    } else {
                        permissions = null
                    }

                    resultSet.close()

                    return@thenApply permissions
                }
            } catch (e: SQLException) {
                e.printStackTrace()
                return@thenApply null
            }
        }
    }

    fun delete(player: UUID) {
        this.clearCache(player)

        this.mySQL.prepareAsync("DELETE FROM permissions WHERE player = ?").whenComplete { statement: PreparedStatement, _: Throwable? ->
            try {
                statement.use {
                    statement.setString(1, player.toString())
                    statement.execute()
                }
            } catch (e: SQLException) {
                e.printStackTrace()
            }
        }
    }

    fun clearCache(player: UUID) {
        this.userPermissions.fastRemoveAsync(player)
    }

    fun cachedPermissions(player: UUID): Set<String>? {
        return this.userPermissions.getOrDefault(player, HashSet())
    }

    fun hasCachedPerm(player: UUID, perm: String): Boolean {
        return this.userPermissions.getOrDefault(player, HashSet()).contains(perm)
    }
}